typing/typecore.ml

```diff
@@ -5742,8 +5737,14 @@ let report_error ~loc env = function
   | Apply_non_function { funct; func_ty; previous_arg_loc; extra_arg_loc } ->
       begin match get_desc func_ty with
         Tarrow _ ->
+          let rec res_type_is_unit ty = match get_desc ty with
```

Comment:
The location of my comment was probably confusing but I would propose to move the expansion call just before the pattern match:
```ocaml
match get_desc (expand_head env rhs) with
```
rather than before the recursive call to `res_type_is_unit` to avoid any question of the form "was the type expanded enough before pattern matching"?

On the very nitpicking side, the responsibility of computing the return type of the function could be left to `report_too_many_arg_error` by transferring  the current env as an argument, but the current location also makes sense.

Reply:
I rewrote it with a different approach that doesn't have this problem.

Reply:
I removed this assumption. It will be useful if this code is ever copy-pasted or if the code around is changed.

I like that it's not in `report_too_many_arg_error` as it shows what information is needed and doesn't make assumptions.